### PR TITLE
[SPARK-16771][SQL] WITH clause should not fall into infinite loop.

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR changes the CTE resolving rule to use only **forward-declared** tables in order to prevent infinite loops. More specifically, new logic is like the following.
- Resolve CTEs in `WITH` clauses first before replacing the main SQL body.
- When resolving CTEs, only forward-declared CTEs or base tables are referenced.
  - Self-referencing is not allowed any more.
  - Cross-referencing is not allowed any more.

**Reported Error Scenarios**

``` scala
scala> sql("WITH t AS (SELECT 1 FROM t) SELECT * FROM t")
java.lang.StackOverflowError
...
scala> sql("WITH t1 AS (SELECT * FROM t2), t2 AS (SELECT 2 FROM t1) SELECT * FROM t1, t2")
java.lang.StackOverflowError
...
```

Note that `t`, `t1`, and `t2` are not declared in database. Spark falls into infinite loops before resolving table names.
## How was this patch tested?

Pass the Jenkins tests with new two testcases.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @rxin .
Could you review this PR?

**Comment by hvanhovell** [comment]
@dongjoon-hyun I think this has merit. I do have one question, what do other databases do? Like postgresql, mysql, sqlserver and others?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -131,17 +131,17 @@ class Analyzer(
...
-        // hive will use the table in database, not the CTE one.
-        // Taking into account the reasonableness and the implementation complexity,
-        // here use the CTE definition first, check table name only and ignore database name
-        // see https://github.com/apache/spark/pull/4929#discussion_r27186638 for more info
         case u : UnresolvedRelation =>
-          val substituted = cteRelations.get(u.tableIdentifier.table).map { relation =>
-            val withAlias = u.alias.map(SubqueryAlias(_, relation))
-            withAlias.getOrElse(relation)
+          try {
+            catalog.lookupRelation(u.tableIdentifier, u.alias)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: This integrates some of the functionality of the `ResolveRelations` rule, but not all (for instance file based datasources). Shouldn't we be consistent and (perhaps) integrate these rules?

**Comment by dongjoon-hyun** [comment]
Thank you for review, @hvanhovell .

For the recursive CTE queries, traditional DBMS supports optional `RECURSIVE` keyword which computing **closures**, i.e., it iterates the subquery until the result converge. For example, PostgreSQL, MSSQL, SQLite supports `WITH RECURSIVE` syntax officially.

For the overlap of `ResolveRelations` rule, this rule calls `lookupRelation` to check the name conflict between the base relations and CTEs. As a side effect, it resolves the relations, too. Here is the reasons.
  - To check the conflict, this is the only simplest way.
  - For the performance, this patch only call one `lookupRelation` for `UnresolvedRelation`.
    - `ResolveRelations` rule do not try resolve again and will not complain about that.
  - The remaining general and exceptional rules like file-based datasource should be in `ResolveRelations`. We can add more datasources there, not here.

For Hive, the recursive queries and `RECURSIVE` syntax are not supported.

**Comment by hvanhovell** [comment]
@dongjoon-hyun 
#### New behavior versus existing systems
I was not talking about recursive CTE's (which can be very useful in some cases). We are changing the behavior of the Analyzer, and this might surprise users. I would like to know if there is a common approach to this among other systems; so we can justify the change in behavior.

#### Resolve Relations
The change you propose is almost subsuming the `ResolveRelations` rule. I think it is correct to to do this. However I am not a big fan of duplicating the behavior in two rules. Do you think there is way to get around this? Could we for instance centralize the logic in `ResolveRelations`?

**Comment by dongjoon-hyun** [comment]
Oh, sorry for misunderstanding, @hvanhovell . I think we can integrate `ResolveRelation` and this one. I'll be back after some testing in this evening. Also for the common approach, I will bring some  references, too.

**Comment by dongjoon-hyun** [comment]
Hi, @hvanhovell . It seems not clearly documented, so I did some comparisons.

First of all, I found that I did overlooked the behavior of the Hive queries. Hive also uses CTE names first in the consecutive CTE queries like other DBMS. It's natural.
```
with t as (select 10), s as (select * from t) select * from s;
```

For the self recursion, the approaches are different. Hive/Oracle raises exceptions, PostgreSQL uses the base tables.
```
with t as (select 10 from t), s as (select * from t) select * from s;
```

The cross referencing case raises exceptions.
```
with t as (select 10 from s), s as (select * from t) select * from s;
```

To sum up, the general approach seems to
- use **forward-declared(not self)** base table or CTE name
- prevent the execution of cyclic queries by raising exceptions.

The root cause of previous Spark behavior is using `self` or `backward-declared CTE` for table name resolutions. I will try to revise this PR according to the above rules. Please comment about the above rule.

**Comment by dongjoon-hyun** [comment]
The above approach also can remove the duplicated scope issue between `CTESubstitution` and `ResolveRelation`.

**Comment by dongjoon-hyun** [comment]
Hi, @hvanhovell .
Sorry for late update. I updated the PR description and code.
Could you review this PR again?

**Comment by dongjoon-hyun** [comment]
Could you review this PR again, @hvanhovell ?
Last time, I did a huge mistake, but now I think I fixed correct according to your advice.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -125,21 +125,23 @@ class Analyzer(
...
     // TODO allow subquery to define CTE
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators  {
-      case With(child, relations) => substituteCTE(child, relations)
+      case With(child, relations) if relations.nonEmpty =>
+        var resolvedRelations = Seq(relations.head._1 -> ResolveRelations(relations.head._2))
+        relations.tail.foreach { case (name, relation) =>
+          resolvedRelations = resolvedRelations ++
+            Seq(name -> ResolveRelations(substituteCTE(relation, resolvedRelations)))
+        }
+        substituteCTE(child, resolvedRelations)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Here, the CTEs of `WITH` clause are resolved sequentially before applying to main SQL body.

**Comment by hvanhovell** [comment]
@dongjoon-hyun I'll take a look in the morning (CET)

**Comment by dongjoon-hyun** [comment]
Thank you so much! Then, see you later. :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -125,21 +125,23 @@ class Analyzer(
   object CTESubstitution extends Rule[LogicalPlan] {
     // TODO allow subquery to define CTE
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators  {
-      case With(child, relations) => substituteCTE(child, relations)
+      case With(child, relations) if relations.nonEmpty =>
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: `relations` is a map. The iteration sequence of a map does not need the be the ordered in which elements were added (Use a map larger than 4 to see this in action, e.g.: `Seq.tabulate(5)(i => ('a' + i).toChar.toString -> i).toMap`).

So I think we need to change the data type in With to accommodate this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -125,21 +125,23 @@ class Analyzer(
...
+        var resolvedRelations = Seq(relations.head._1 -> ResolveRelations(relations.head._2))
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Entire Block: Either use a truly imperative approach or use foldLeft(...). This is a bit of both.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: So, I changed it to `Seq`. Now, `relations` of `With` is a sequence.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 5]
```diff
@@ -392,11 +392,9 @@ case class InsertIntoTable(
  * This operator will be removed during analysis and the relations will be substituted into child.
  *
  * @param child The final query of this CTE.
- * @param cteRelations Queries that this CTE defined,
- *                     key is the alias of the CTE definition,
- *                     value is the CTE definition.
+ * @param cteRelations A sequence of pair (alias, the CTE definition) that this CTE defined
  */
-case class With(child: LogicalPlan, cteRelations: Map[String, SubqueryAlias]) extends UnaryNode {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Here~

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -125,21 +125,23 @@ class Analyzer(
...
-        // Taking into account the reasonableness and the implementation complexity,
-        // here use the CTE definition first, check table name only and ignore database name
-        // see https://github.com/apache/spark/pull/4929#discussion_r27186638 for more info
+    def substituteCTE(plan: LogicalPlan, cteRelations: Seq[(String, LogicalPlan)]): LogicalPlan = {
+      plan transformDown {
         case u : UnresolvedRelation =>
-          val substituted = cteRelations.get(u.tableIdentifier.table).map { relation =>
-            val withAlias = u.alias.map(SubqueryAlias(_, relation))
-            withAlias.getOrElse(relation)
+          val substituted = cteRelations.find(_._1 == u.tableIdentifier.table).map(_._2).map {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Is this use case sensitive resolution? In this case it would be useful if cteRelations *is* a map.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Never mind you have already done that :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 7]
```diff
@@ -392,11 +392,9 @@ case class InsertIntoTable(
  * This operator will be removed during analysis and the relations will be substituted into child.
  *
  * @param child The final query of this CTE.
- * @param cteRelations Queries that this CTE defined,
- *                     key is the alias of the CTE definition,
- *                     value is the CTE definition.
+ * @param cteRelations A sequence of pair (alias, the CTE definition) that this CTE defined
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Perhaps document that the CTE should be ordered in the way that they are defined.

Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Yep. I'll add that clearly.

**Comment by hvanhovell** [comment]
@dongjoon-hyun this is looking very promising. I left two small comments.

**Comment by dongjoon-hyun** [comment]
Thank you, @hvanhovell !

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: This does not replace `With` in all cases. Could you remove the `relations.nonEmpty` guard?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 8]
```diff
@@ -125,22 +125,20 @@ class Analyzer(
   object CTESubstitution extends Rule[LogicalPlan] {
     // TODO allow subquery to define CTE
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators  {
-      case With(child, relations) => substituteCTE(child, relations)
+      case With(child, relations) if relations.nonEmpty =>
+        substituteCTE(child, relations.foldLeft(Seq.empty[(String, LogicalPlan)])((resolved, r) =>
+          resolved ++ Seq(r._1 -> ResolveRelations(substituteCTE(r._2, resolved)))))
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: `resolved :+ r._1 -> ResolveRelations(substituteCTE(r._2, resolved))`?

You could also deconstruct the `r` tuple for easier reading.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: If WITH without relations, I thought we can skip all since there is nothing to be replaced.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Yeah you are right about that. But the current code does not remove the `With` node if there are no relations.

Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Oh, I see!

**Comment by hvanhovell** [comment]
LGTM pending Jenkins.

**Comment by dongjoon-hyun** [comment]
Rebased just to resolve conflicts.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 9]
```diff
@@ -2938,4 +2938,26 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
         data.selectExpr("`part.col1`", "`col.1`"))
     }
   }
+
+  test("SPARK-16771: WITH clause should not fall into infinite loop") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: can you create a test in SQLQueryTestSuite instead?

Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Sure. I'll move this.

Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Ur, @rxin .
`SQLQueryTestSuite` seems not to support exceptions cases yet.

Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: In this PR, `exception`s are important and should be checked.

**Comment by dongjoon-hyun** [comment]
Oh, now `Exception` is supported at https://github.com/apache/spark/commit/0db373aaf87991207a7a8a09853b6fa602f0f45b. Thanks, @rxin . I will try again.

**Comment by dongjoon-hyun** [comment]
Hi, @rxin . 
Now the testcases are moved into `sql-tests` with `SQLQueryTestSuite`.

**Comment by dongjoon-hyun** [comment]
Hi, @rxin . I moved the testcase into new suite.
Please let me know if there is more thing to do.
New SQL test suite is great!

File: sql/core/src/test/resources/sql-tests/inputs/with.sql [hunk 10]
```diff
@@ -0,0 +1,14 @@
+create temporary view t as select * from values 0, 1, 2 as t(id);
```
Comment (rxin) [inline sql/core/src/test/resources/sql-tests/inputs/with.sql]: maybe CTE instead of "with" ?

Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/inputs/with.sql]: Here, t1 and t2 is base tables.
The followings are used CTEs and to check if base table or previous CTE is used correctly.

Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/inputs/with.sql]: Ah, I see. You mean `with.sql` into `CTE.sql`.
I see. No problem.

Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/inputs/with.sql]: I will use `cte.sql` instead of `with.sql`.

**Comment by dongjoon-hyun** [comment]
Thank you, @rxin . It's updated.

**Comment by dongjoon-hyun** [comment]
Hi, @hvanhovell .
Could you review this again? After the last time you've seen, only the testcase is moved into a new testsuite.

**Comment by hvanhovell** [comment]
LGTM (again) - merging to master. Thanks!

**Comment by dongjoon-hyun** [comment]
Thank you, @hvanhovell !
